src/errors/MetaplexError.ts

```diff
@@ -0,0 +1,66 @@
+export type MetaplexErrorSource = 'sdk' | 'network' | 'rpc' | 'plugin' | 'program';
+
+export type MetaplexErrorInput = {
+  key: string;
+  title: string;
+  problem: string;
+  solution: string;
+  source: MetaplexErrorSource;
+  sourceDetails?: string;
+  cause?: Error;
+};
+
+export type MetaplexErrorInputWithoutSource = Omit<MetaplexErrorInput, 'source' | 'sourceDetails'>;
+
+export class MetaplexError extends Error {
+  readonly name: 'MetaplexError' = 'MetaplexError';
+  readonly key: string;
+  readonly title: string;
+  readonly problem: string;
+  readonly solution: string;
+  readonly source: MetaplexErrorSource;
+  readonly sourceDetails?: string;
+  readonly cause?: Error;
+
+  constructor(input: MetaplexErrorInput) {
+    super(input.problem);
+    this.key = `metaplex.errors.${input.key}`;
+    this.title = input.title;
+    this.problem = input.problem;
+    this.solution = input.solution;
+    this.source = input.source;
+    this.sourceDetails = input.sourceDetails;
+    this.cause = input.cause;
+    this.message = this.toString(false);
+  }
+
+  getCapitalizedSource(): string {
+    if (this.source === 'sdk' || this.source === 'rpc') {
+      return this.source.toUpperCase();
+    }
+
+    return this.source[0].toUpperCase() + this.source.slice(1);
+  }
+
+  getFullSource(): string {
+    const capitalizedSource = this.getCapitalizedSource();
+    const sourceDetails = this.sourceDetails ? ` > ${this.sourceDetails}` : '';
+
+    return capitalizedSource + sourceDetails;
+  }
+
+  toString(withName: boolean = true) {
```

Comment:
Not sure about this `toString` thingy since it may not print to the console as one would expect when the error goes uncaught, i.e. have you tried `throw new WhateverError(..)` and left it uncaught?
I'm not sure it'll print this, but actually the `message + stack` or similar.

Reply:
I have tested this in a [TypeScript Play](https://www.typescriptlang.org/play?#code/C4TwDgpgBAshwEMwBsIA8CiAnLB7LAyrgK5YDG0AvFAOQDOAJgNY1QA+tAdvAO74vtaWMGVYcaKYgHMAlpzG0weKVgQBbGgG4AUNtCRY8JKkw58ASU5hiwKNQDe2qFCYQQALih1gWOVIA0TlDAMsCont6+nAFBSrgARqhqET5+gc50uMg2MricKVExGSTkEJ5wiCjo2HiEJRTpXvUQACJGMsh0APwFaUFkCMR0ED1QNfiBAL66+tAVxtVmWJbWwADqoQAWJMBEpBR2UADyaqEAPPNVprUrNv60mfsQCvTNbYgddDQAfDraZMgEHQ6IZKiZxlgoOhgBBOAwQRCoI5nFgIAgGHlkCAoJx1GVaJdwUtWNQaITFrUtEFUejMdjXB4vKlojoUWiMZwscFQuEmYVWVAaRyuXFEhBkny-AKhXSmtkQnleizqezZY9SuUjFcIXtStLVZzseqKO8EJ9RpEpSraYaoAMhiNPBC-s4yHlIsQyMB8AAKOSrTVgikWKw2ACUSKCGWIkCwftDwAAdKKkmGBc5gJsZHREwzDgADNRakyJiBLHMAEns-psubck3z6eCWZzITCVCgNaTbdQTcz2eTeDFakOXcHCSSfZbicy8tynFHCZnWRyeSnA+NHbHm-XOc3ps+i9Wy6eB86u8T9uGR9rV4gF6LwIQUg7-dbuAIzKkPoAZghOhAaZBNMQQvsAADCSChP+MgAF4QAwuoUD6YZKlIkbOM4Mg-lAPpviepR2JQpKMAIbAcPhm5EaSwiiBGyKYWywCkAulHNIm3oAKpgLGkHDChTYgYxqLMVgrHTpuADaAAMAC6HG4NxvFAhAKFQAA1M2G7sXQyAyMhACMQHOEJUBgQAYsQyDIEhqmoZK0QYZhbqcN4dpQYgenwYhzSHPhYGQWA0FeQhtkCVGdrurY+7tJ0fkSW8sUgl0UD5lA3xQFWbGnklDZQJ4NBUhFIkse5QWeXBoW+ZpMUfOewG6BmH5fj6PBbAAcninjxLgWRogu1A+MQgFOa6UXuQ6DAAELYoN053lAKX5gAOpwq2QZNUAzZ4WXzYMwx5QVRXCfApU+hFzitR1eKLalkm7QOuJFpMsmpfltA0BG6kXalD2tjyEB5d9jHOCtnDfBltk7fY-nwJZ1lhWGQM-WDENQAACkOSTQ-hKbisjIOpataNEHOiqZTDCVk5wBMg6jGUANJuDj04MrTjF3tN2LAyDNCrTQEXGVA0wgQCQIglNxBwsgWCItCsLwqCCzXPgTkuR6Xq+l2gbKxCtzrFsOy2fREV0DGZY+gxjGJjbXaNIxDKePmkiyJwibxFLDAy4mVZjmz+b25hPb4vmkvS5CGW+0uwcNoHxRPAVLtyDQcdNDldV0AVYde1gKcRZMQumd4CAhGQnaucQP4-vpMiwsAllwnQPrWIk+lM4ylrRPcd6jBCJsnaJC7cDwWlN1bzn7RAqeO7Qchm1XNd1wA+j+ntfKnwcFZY8-V2QtecLYDfwnnhN4xKoeezLkVWQwOK4LYq9wlAQx+M20BxAAbjIDAIVAPD-qgWw917Atz0mQduL1EwBx+rOVc+RaDo1QCpKA6Jb5qHwNAAgRwAAyIJvRQBACUP+AD4AoKfj4bEz4zRuxPphAuAoi6IFLnfYAABNeA5g1BVCLAfBCPoe5OiWP3TCJUxI4ggCPN8lsfp3mnszLg98l4gHgEvGQXCkh1wQrQxim9aDtXvlAdhthOHcM0QwbRmEz4FQACotigD+NEoloDZhYQQ0hajTG8IYImCxxRqYFUQWia8ZBNgQDIEwKA8QEDhKgICGEWAfGB3oQ1ECFCnKZjwCPbOMsISJjnpXXe+965rx9K8IsS9QH6SXgyFOW1L6yyWImTg98jEmI0V4lCQFJjuWACE3CI0GLYX6eXYunAKC4BwuSFWWBhFjVcn1RMyBcDfggDORA4TC7aGmEAA). 🙂